doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+UserWarning: Provided arguments did not match declared signature, recompiling...
+Tracing occurring
+array(0.16996714)
+
+Try and compile the full workflow
+---------------------------------
+
+When porting your PennyLane code to work with Catalyst and :func:`@qjit <.qjit>`, the
+biggest performance advantage you will see is if you qjit
+your *entire* workflow, not just the QNodes. So think about putting
+everything inside your JIT-compiled function, including for loops
+(including optimization loops), gradient calls, etc.
+
+Consider the following PennyLane example, where we have a parametrized
+circuit, are measuring an expectation value, and are optimizing the result:
+
+.. code-block:: python
+
+    dev = qml.device("default.qubit", wires=4)
+
+    @qml.qnode(dev)
+    def cost(weights, data):
+        qml.AngleEmbedding(data, wires=range(4))
+
+        for x in weights:
+            # each trainable layer
+            for i in range(4):
+                # for each wire
+                if x[i] > 0:
+                    qml.RX(x[i], wires=i)
+                elif x[i] < 0:
+                    qml.RY(x[i], wires=i)
+
+            for i in range(4):
+                qml.CNOT(wires=[i, (i + 1) % 4])
+
+        return qml.expval(qml.PauliZ(0) + qml.PauliZ(3))
+
+    weights = jnp.array(2 * np.random.random([5, 4]) - 1)
+    data = jnp.array(np.random.random([4]))
+
+    opt = jaxopt.GradientDescent(cost, stepsize=0.4)
+
+    params = weights
+    state = opt.init_state(params)
+
+    for i in range(200):
+        (params, _) = tuple(opt.update(params, state, data))
+
+Using PennyLane v0.32 on Google Colab with the Python 3 Google Compute Engine
+backend, this optimization takes 3min 28s ± 2.05s to complete.
+
+Let's switch over to Lightning, our high-performance statevector simulator,
+alongside the adjoint differentiation method. To do so, we change the first
+two lines of the above code-block to set the device as ``"lightning.qubit"``,
+and specify ``diff_method="adjoint"`` in the QNode decorator. With this
+change, we have reduced the execution time down to 30.7s ± 1.8s.
+
+We can rewrite this QNode to use Catalyst control flow, and compile
+it using Catalyst:
+
+.. code-block:: python
+
+    dev = qml.device("lightning.qubit", wires=4)
+
+    @qjit
+    @qml.qnode(dev)
+    def cost(weights, data):
+        qml.AngleEmbedding(data, wires=range(4))
+
+        def layer_loop(i):
+            x = weights[i]
+            def wire_loop(j):
+
+                @cond(x[j] > 0)
+                def trainable_gate():
+                    qml.RX(x[j], wires=j)
+
+                @trainable_gate.else_if(x[j] < 0)
+                def negative_gate():
+                    qml.RY(x[j], wires=j)
+
+                trainable_gate.otherwise(lambda: None)
+                trainable_gate()
+
+            def cnot_loop(j):
+                qml.CNOT(wires=[j, jnp.mod((j + 1), 4)])
+
+            for_loop(0, 4, 1)(wire_loop)()
```

Comment:
I think it's good that unnecessary library imports are removed. But, `cond`, `for_loop`, and `grad` in these examples may be a bit confusing in particular for users new to the control-flow support in Catalyst. What about adding `from catalyst import ...` to these examples?

Reply:
Hmm I might leave this for now, since it matches the convention in our docstrings. But good point, we should revisit this.